x/auth/vesting/types/period.go

```diff
@@ -26,3 +27,267 @@ func (vp Periods) String() string {
 	return strings.TrimSpace(fmt.Sprintf(`Vesting Periods:
 		%s`, strings.Join(periodsListString, ", ")))
 }
+
+// A "schedule" is an increasing step function of Coins over time.
+// It's specified as an absolute start time and a sequence of relative
+// periods, with each step at the end of a period. A schedule may also
+// give the time and total value at the last step, which can speed
+// evaluation of the step function after the last step.
+
+// ReadSchedule returns the value of a schedule at readTime.
+func ReadSchedule(startTime, endTime int64, periods []Period, totalCoins sdk.Coins, readTime int64) sdk.Coins {
+	if readTime <= startTime {
+		return sdk.NewCoins()
+	}
+	if readTime >= endTime {
+		return totalCoins
+	}
+
+	coins := sdk.NewCoins() // sum of amounts for events before readTime
+	time := startTime
+
+	for _, period := range periods {
+		if readTime < time+period.Length {
+			// we're reading before the next event
+			break
+		}
+		coins = coins.Add(period.Amount...)
+		time += period.Length
+	}
+
+	return coins
+}
+
+// max64 returns the maximum of its inputs.
+func max64(i, j int64) int64 {
+	if i > j {
+		return i
+	}
+	return j
+}
+
+// min64 returns the minimum of its inputs.
+func min64(i, j int64) int64 {
+	if i < j {
+		return i
+	}
+	return j
+}
+
+// coinsMin returns the minimum of its inputs for all denominations.
+func coinsMin(a, b sdk.Coins) sdk.Coins {
```

Comment:
nit I think this should be a func of the Coins type and should be split and upstreamed to the SDK

`func (c Coins) Min(other Coins) Coins`

Reply:
Will do when I get a chance. `Coins` also needs an equality test that doesn't panic. I've got some work in progress for `agoric-sdk` that adds some utilities with tests for their algebraic properties: https://github.com/Agoric/agoric-sdk/blob/3991-standalone-loc/golang/cosmos/types/coin.go

Reply:
can you create an issue?

Reply:
Created https://github.com/cosmos/cosmos-sdk/issues/10995

Please assign me if you wish - it's not letting me assign myself.